### ISSUE TITLE
Escape quotes in filename used as start

### ISSUE THREAD

**Issue by GrumpyLittleTed** [oven-sh/bun#437]
Issue is that a path containing a single quote is not escaped properly when imported as start

cd /tmp/
mkdir "joe's stuff" 
cd "joe's stuff"
echo "console.log(99)" >z.js
bun z.js

Error is
error: Expected ";" but found "s"

import * as start from '/tmp/joe's stuff/z.js';
                                 ^
bun:main:3:34 95

**Comment by Pruxis** [comment]
It seems like some characters break the import code.
Not only ' but also \, and probably some others too.

**Comment by Jarred-Sumner** [comment]
the bug is probably somewhere in here

https://github.com/Jarred-Sumner/bun/blob/41f4302d394a6d2696a2f7b3a6ab98474270bfa5/src/js_parser.zig#L8961-L9000

Edit: your sytnax above is incorrect

>  '/tmp/joe's stuff/z.js';

you need to escape `'` in `joe's`

**Comment by rabingaire** [comment]
@Pruxis I was actually looking to fix this bug, can you let me know which OS were you using when the comma broke the import code.

**Comment by GrumpyLittleTed** [comment]
I've had a look at the Zig code, however I'm _guessing_ that the bug is not in the parsing code but instead in the code that generates the following line:
**import * as start from '/tmp/joe's stuff/z.js';**
where the stuff between the surrounding single quotes needs to be escaped on output (escaping first the quote character \ and then the terminating character ' ).

Incidentally I discovered this when using a symlink:

ln -sfn /tmp/joe\'s\ stuff/ slink
cd slink && bun z.js

The generated code uses the canonicalised path (what you'd get with readlink -f z.js) so in some cases working around this bug would be quite awkward (hard links are a bit limited here if the files are spread across different file systems).

**Comment by rabingaire** [comment]
Yes, I also found the same, my solution was also to try to escape invalid chars from the string. I did a bit of research and there are so many cases that need to be handled you can check this out if you are interested in more details:

https://stackoverflow.com/questions/1976007/what-characters-are-forbidden-in-windows-and-linux-directory-names

**Comment by GrumpyLittleTed** [comment]
My example was run on Linux. Isn't it "just" an output escaping issue though (previous comment) so the two characters to escape regardless: the escape character itself and the terminator which matches the type of quote used to surround the string. Of course non-printable characters need to be converted to a format like \x too.

I've not looked at the bun code so I'm assuming that it, somehow, generates the import from start stuff, and this comment is based on that assumption.

**Comment by GrumpyLittleTed** [comment]
I've had a quick poke around in the code now. My best guess is that 

a) culprit for this bug is entry_points.zig
b) in this code both the dir_to_use and original_path.filename need to be escaped so that 
    1) any '\' become '\\' 
    2) single quotes become \'

Actually Typescript (tsc) works fine with dir names that contain single quotes, but appears to fail to find the filename (basename) when it contains a quote, so #1 above really only currently applies to the dir name component.
    
I'm not familiar enough with Zig to supply a fix. 

Current  zigcode:
        const dir_to_use: string = original_path.dirWithTrailingSlash();
        const code = try std.fmt.bufPrint(
            &entry.code_buffer,
            \\//Auto-generated file
            \\var cjsSymbol = Symbol.for("CommonJS");
            \\import * as start from '{s}{s}';
            \\export * from '{s}{s}';
            \\var entryNamespace = start;
            \\var cjs = start?.default;
            \\if (cjs && typeof cjs ===  'function' && cjsSymbol in cjs) {{
            \\  entryNamespace = cjs();
            \\}}
            \\if (typeof entryNamespace?.then === 'function') {{
            \\   entryNamespace = entryNamespace.then((entryNamespace) => {{
            \\      if(typeof entryNamespace?.default?.fetch === 'function')  {{
            \\        Bun.serve(entryNamespace.default);
            \\      }}
            \\   }}, reportError);
            \\}} else if (typeof entryNamespace?.default?.fetch === 'function') {{
            \\   Bun.serve(entryNamespace.default);
            \\}}
            \\
        ,
            .{
                dir_to_use,
                original_path.filename,
                dir_to_use,
                original_path.filename,
            },
        );

**Comment by Electroid** [comment]
It looks like this has been fixed as of `0.2.2`, tested on MacOS Silicon. If this is not the case, please comment and we'll re-open the issue.

```js
// joe's stuff.js
export function stuff() {
  return "stuff";
}
```

```js
// index.ts
import { stuff } from "./joe's stuff";
console.log(stuff()); // => "stuff"
```
